develop/tutorials/articles/310-troubleshooting/14-jndi-service-builder.markdown

```diff
@@ -0,0 +1,158 @@
+   	            <column name="id" db-name="id" primary="true" type="long" />
+   	            <column name="foo" db-name="foo" type="String" />
+   	            <column name="bar" db-name="bar" type="long" />
+            </entity>
+        </service-builder>
+
+    Note the following `<entity>` tag attributes from this example: 
+
+    - `data-source`: The data source's name. You can choose whatever name you 
+    want. You'll create a Spring bean with the same name that points to the JNDI 
+    data source you want to use. 
+    - `table`: The name of the database table that corresponds to your entity. 
+
+    Also note that in each of your entity's `<column> tags`, you must set the 
+    `db-name` attribute to the name of that entity's database column. 
+
+2.  Create a Spring bean that points to the JNDI data source you want to use. To 
+    do this, create a file called `ext-spring.xml` in your Service Builder 
+    module's `src/main/resources/spring` folder. In this file, define the 
+    following: 
+
+    - A Spring bean for the JNDI data source you want to use. 
+    - A @product@ data source bean that points to your JNDI data source bean. 
+    - An alias for the @product@ data source bean. 
+
+    Here's an example: 
+
+        <?xml version="1.0" encoding="UTF-8"?>
+
+        <beans default-destroy-method="destroy" default-init-method="afterPropertiesSet"
+            xmlns="http://www.springframework.org/schema/beans" 
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://www.springframework.org/schema/beans 
+                    http://www.springframework.org/schema/beans/spring-beans-3.0.xsd">
+
+            <bean class="com.liferay.portal.dao.jdbc.spring.DataSourceFactoryBean"
+   	            id="jndiDatasource">
+   	            <property name="propertyPrefix" value="custom." />
+   	            <property name="properties">
+                    <props>
+                        <prop key="custom.jndi.name">jdbc/externalDataSource</prop>
+                    </props>
+   	            </property>
+            </bean>
+
+            <bean class="org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy"
+                id="liferayDataSource">
+   	            <property name="targetDataSource" ref="jndiDatasource" />
+            </bean>
+
+            <alias alias="extDataSource" name="liferayDataSource" />
+
+        </beans>
+
+3.  Run Service Builder. To do this via the command line for a Gradle project, 
+    navigate to your Service Builder module project's root folder and run 
+    `gradlew buildService`. Note that you can also run Service Builder from 
+    within Liferay @ide@. For more information on this, see 
+    [the tutorial on running Service Builder](/develop/tutorials/-/knowledge_base/7-0/running-service-builder-and-understanding-the-generated-code). 
+
+4.  Now you can access the JNDI data source and use it however you wish. Here's 
```

Comment:
There's a strange jump here. First, you configure it with Service Builder and generate services. The next logical step then should be using those services Service Builder generated. Instead, you jump to using the data source directly (i.e., not through Service Builder) and executing your own SQL queries. We should discuss this before we publish.